packages/language-server/src/linting.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  findEndPosition,
+  parserWrapper,
+  validateSyntax,
+} from '@neo4j-cypher/language-support';
+import debounce from 'lodash.debounce';
+import { join } from 'path';
+import { Diagnostic, TextDocumentChangeEvent } from 'vscode-languageserver';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import workerpool from 'workerpool';
+import { LinterTask, LintWorker } from './lint-worker';
+
+const pool = workerpool.pool(join(__dirname, 'lint-worker.js'), {
+  minWorkers: 2,
+});
+
+let lastSemanticJob: LinterTask | undefined;
+
+async function rawLintDocument(
+  change: TextDocumentChangeEvent<TextDocument>,
+  sendDiagnostics: (diagnostics: Diagnostic[]) => void,
+) {
+  const { document } = change;
+
+  const query = document.getText();
+  if (query.length === 0) {
+    return;
+  }
+
+  const syntaxErrors = validateSyntax(query, {});
+
+  sendDiagnostics(syntaxErrors);
+
+  if (syntaxErrors.length === 0) {
+    try {
+      if (lastSemanticJob !== undefined && !lastSemanticJob.resolved) {
+        void lastSemanticJob.cancel();
+      }
+
+      const proxyWorker = (await pool.proxy()) as unknown as LintWorker;
+      lastSemanticJob = proxyWorker.validateSemantics(query);
```

Comment:
could pass positions here to avoid doing the the findposition in the main thread